src/Services/Container.php

```diff
@@ -73,14 +89,74 @@ public function loadService(string $name, bool $exception = TRUE) {
    *   Arguments that should be passed to the hook invocations.
    */
   public function loadSpecsFromHook(string $hook, ...$arguments) {
-    $specs = module_invoke_all($hook, ...$arguments);
-    foreach ($specs as &$spec) {
-      if (!is_array($spec)) {
-        $spec = ['class' => $spec];
-      }
-    }
+    $info = module_invoke_all($hook, ...$arguments);
+    $specs = $this->processInfo($info);
     drupal_alter($hook, $specs, ...$arguments);
-    $this->specs += $specs;
+    $this->specs = array_replace($this->specs, $specs);
```

Comment:
What's the advantage of `array_replace()` here?

Reply:
It does something different. If you disregard the write operations and the order of keys then an equivalent is: 

```php
$this->specs = $specs + $this->specs;
```

This is the exact opposite of the replaced line.

---

A more precise replacement would be:

```php
foreach ($specs as $key => $value) {
  $this->specs[$key] = $value;
}
```

This keeps the order of keys in `$this->specs`.

Reply:
Thanks! Does the order of keys matter?

Reply:
No :-) I only came to think of `$this->specs` as being many and `$specs` being few. So:

```php
array_replace($this->specs, $specs);
```

seemed more efficient:

```php
$specs + $this->specs;
```

Happy to change it around though if it helps readability.

Reply:
No, totally fine with array_replace(), was just wondering why you preferred it ;)